### PR TITLE
Download file to specific destination

### LINKED ISSUE #1240: Opt-in for downloading without symlinks
This is possibly a niche use case.

I recently found that some libraries (`coremltools`, in this case) don't play nice with symlinks even on Unix platforms 😲. This led me to replace this one-liner, which was intended for user communication:

```Python
from huggingface_hub import snapshot_download

repo_id = "apple/coreml-stable-diffusion-v1-4"
variant = "original/packages"

downloaded = snapshot_download(repo_id, allow_patterns=f"{variant}/*")
```

With this one (taken from [the blog post](https://huggingface.co/blog/diffusers-coreml#download-the-model-checkpoints)):

```Python
from huggingface_hub import snapshot_download
from huggingface_hub.file_download import repo_folder_name
from pathlib import Path
import shutil

repo_id = "apple/coreml-stable-diffusion-v1-4"
variant = "original/packages"

def download_model(repo_id, variant, output_dir):
    destination = Path(output_dir) / (repo_id.split("/")[-1] + "_" + variant.replace("/", "_"))
    if destination.exists():
        raise Exception(f"Model already exists at {destination}")
    
    # Download and copy without symlinks
    downloaded = snapshot_download(repo_id, allow_patterns=f"{variant}/*", cache_dir=output_dir)
    downloaded_bundle = Path(downloaded) / variant
    shutil.copytree(downloaded_bundle, destination)

    # Remove all downloaded files
    cache_folder = Path(output_dir) / repo_folder_name(repo_id=repo_id, repo_type="model")
    shutil.rmtree(cache_folder)
    return destination

model_path = download_model(repo_id, variant, output_dir="./models")
print(f"Model downloaded at {model_path}")
```

It's not the end of the world, but in this case I _really_ wanted to stress how easy it was to download Core ML checkpoints from the hub and use them downstream for whatever purpose.

If this is something that only affects `coremltools`, then it's not worthwhile doing anything (I'll open a PR there when I look into the problem in more depth). I'm raising the issue in case somebody else has observed other use cases that could benefit from a flag to unconditionally use #1067 even if symlinks are supported by the underlying os.

### PR DESCRIPTION
**Pull request by Wauplin**
Resolve #1240 (see discussion there) following what we agreed on in https://github.com/huggingface/huggingface_hub/issues/1240#issuecomment-1438492986.

I have changed the naming (`local_dir` instead of `destination` for consistency with other methods and `local_dir_use_symlinks` instead of just `use_symlinks` to make it explicit). So to sum up the behavior:

```py
# pip install git+https://github.com/huggingface/huggingface_hub1240-snapshot-download-to-destination
from huggingface_hub import snapshot_download

# Download and cache files
snapshot_download(repo_id)

# Download and cache files + duplicate small files (<5MB) to "my-folder" + add symlinks for big files
snapshot_download(repo_id, local_dir="my-folder")
snapshot_download(repo_id, local_dir="my-folder", local_dir_use_symlinks="auto")

# Download and cache files + add symlinks in "my-folder"
snapshot_download(repo_id, local_dir="my-folder", local_dir_use_symlinks=True)

# Duplicate files already existing in cache
# and/or
# Download missing files directly to "my-folder"
#     => if ran multiple times, files are re-downloaded
snapshot_download(repo_id, local_dir="my-folder", local_dir_use_symlinks=False)
```

@philschmid @pcuenca I would be very interested if you could try this PR and give me some feedback. Also @peterschmidt85 since you showed great interest in this issue.

Additional notes:
- default is `local_dir_use_symlinks="auto"`: big files (>5MB) are symlinked to save disk usage but small files are duplicated to ease user experience. Usually we expect small files to be potentially manually edited by the user. File duplication do not take too much space and ensure the user will not corrupt the cache unexpectedly. Default 5MB threshold can be configured with `HF_HUB_LOCAL_DIR_AUTO_SYMLINK_THRESHOLD` environment variable.
- When setting `local_dir_use_symlinks=False` the cache directory will still be used and potentially have some modifications even if it's empty => I'm talking about the `snapshots/` folder being created (empty), the `refs` folder being created and populated. Those files are neglectable in term of size but still exist.
- If files already exists in `local_dir`, they are overwritten without warning.
- At the moment, if `local_dir_use_symlinks=False`
    - If files are not cached, we don't have any reference kept to which version/reference the file comes from. It means that if a user run the command again, the file will be download again. To solve that we could compute the sha of the file and compare but that's out of scope of this PR I think. Users using `local_dir_use_symlinks=False` should be aware it's not optimal.
    - If files are cached, re-running the command will duplicate again the file (i.e. even if file exists at destination, it will be copy-pasted again). It doesn't have impact on disk usage but it's not optimized computation-wise. Same as above, I'll not fix it in this PR.
- [x] Updated [the "Download files" guide](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1360/en/guides/download):

### CODE REVIEW

**Review by pcuenca — COMMENTED** [review]
I tested this implementation and it works as I expected it to. When merged, it can be used to fully meet the requirements described when I opened #1240.

**Comment by Wauplin** [comment]
Great news! Thank you for testing @pcuenca :)

File: src/huggingface_hub/file_download.py [hunk 1]
```diff
@@ -844,11 +844,19 @@ def _create_relative_symlink(src: str, dst: str, new_blob: bool = False) -> None
...
 
-    cache_dir = os.path.dirname(os.path.commonpath([src, dst]))
-    if are_symlinks_supported(cache_dir=cache_dir):
-        relative_src = os.path.relpath(src, start=os.path.dirname(dst))
+    try:
+        _support_symlinks = are_symlinks_supported(
+            os.path.dirname(os.path.commonpath([os.path.realpath(src), os.path.realpath(dst)]))
+        )
+    except PermissionError:
+        # Permission error means src and dst are not in the same folder (e.g. destination path has been provided
```
Comment (pcuenca) [inline src/huggingface_hub/file_download.py:852]:
```suggestion
        # Permission error means src and dst are not in the same volume (e.g. destination path has been provided
```

(guessing that's what we mean here)

**Review by pcuenca — COMMENTED** [review]
Minor nits.

File: src/huggingface_hub/_snapshot_download.py [hunk 2]
```diff
@@ -42,13 +44,21 @@ def snapshot_download(
...
-    ones you will need a priori. All files are nested inside a folder in order
-    to keep their actual filename relative to that folder.
+    Download a whole snapshot of a repo's files at the specified revision. This is useful when you want all files from
+    a repo, because you don't know which ones you will need a priori. All files are nested inside a folder in order
+    to keep their actual filename relative to that folder. You can also filter which files to download using
+    `allow_patterns` and `ignore_patterns`.
 
-    An alternative would be to just clone a repo but this would require that the
-    user always has git and git-lfs installed, and properly configured.
+    If `local_dir` is provided, the downloaded files will be placed in this location. If `local_dir_use_symlinks` is
```
[the remark this answers is not in the document]
Comment (Wauplin) [reply src/huggingface_hub/_snapshot_download.py:52]: Far better !

**Comment by Wauplin** [comment]
Ping @julien-c @LysandreJik @pcuenca I have updated this PR (and its description) to have `local_folder_use_symlinks="auto"` by default. The idea is to either use symlinks or duplicate files depending on the file size. Small files (<5MB) are duplicated: it doesn't take much space and it ensures users will not manually edit a file in the cache. Big files are still symlinked to preserver disk usage. And in both cases files are cache to preserve bandwidth. I think it's a good trade-off between performance (disk usage/bandwidth) and usability (not corrupting the cache).

File: src/huggingface_hub/file_download.py [hunk 3]
```diff
@@ -963,6 +991,14 @@ def hf_hub_download(
             The version of the library.
         cache_dir (`str`, `Path`, *optional*):
             Path to the folder where cached files are stored.
+        local_dir (`str` or `Path`, *optional*:
```
Comment (LysandreJik) [inline src/huggingface_hub/file_download.py:994]:
```suggestion
        local_dir (`str` or `Path`, *optional*):
```

**Review by LysandreJik — APPROVED** [review]
Impressive PR! That's a cool feature, thanks for working on it @Wauplin

**Pull request merged**